File: analyzers/format_check.py
"""
X-Ray LLM — Format checking, type checking analyzers.
"""

import contextlib
import os
import subprocess

from analyzers._shared import _fwd
from xray.types import FormatResult, TypeCheckResult


def check_format(directory: str) -> FormatResult:
    """Run ruff format --check to find files needing reformatting."""
    try:
        result = subprocess.run(
            ["ruff", "format", "--check", directory],
            capture_output=True,
            text=True,
            timeout=60,
        )
    except FileNotFoundError:
        return {"error": "ruff not found. Install: uv tool install ruff"}
    except subprocess.TimeoutExpired:
        return {"error": "ruff format check timed out."}

    files = []
    for line in result.stdout.strip().split("\n"):
        line = line.strip()
        if line and os.path.isfile(line):
            files.append(_fwd(os.path.relpath(line, directory)))
    for line in result.stderr.strip().split("\n"):
        line = line.strip()
        if line.startswith("Would reformat:"):
            fname = line.replace("Would reformat:", "").strip()
            if fname:
                files.append(_fwd(fname))

    return {
        "needs_format": len(files),
        "files": files[:500],
        "all_formatted": result.returncode == 0,
    }


def check_types(directory: str) -> TypeCheckResult:
    """Run ty type checker on Python files for type-safety diagnostics."""
    try:
        result = subprocess.run(
            ["ty", "check", "--output-format", "concise", directory],
            capture_output=True,
            text=True,
            timeout=120,
        )
    except FileNotFoundError:
        return {"error": "ty not found. Install: uv tool install ty"}
    except subprocess.TimeoutExpired:
        return {"error": "ty type check timed out."}

    diagnostics = []
    raw = (result.stdout + "\n" + result.stderr).strip()
    for line in raw.split("\n"):
        line = line.strip()
        if not line or line.startswith("Found ") or line.startswith("info:"):
            continue
        # Format: file:line:col: severity[rule] message
        parts = line.split(": ", 1)
        if len(parts) >= 2:
            location = parts[0]
            rest = parts[1]
            loc_parts = location.rsplit(":", 2)
            file_path = loc_parts[0] if loc_parts else location
            diag = {
                "file": _fwd(file_path),
                "location": location,
                "message": rest,
            }
            if "error[" in rest:
                diag["severity"] = "error"
            elif "warning[" in rest:
                diag["severity"] = "warning"
            else:
                diag["severity"] = "info"
            diagnostics.append(diag)

    # Extract summary line ("Found N diagnostics")
    total = len(diagnostics)
    for line in raw.split("\n"):
        if line.strip().startswith("Found "):
            with contextlib.suppress(ValueError, IndexError):
                total = int(line.strip().split()[1])
            break

    return {
        "total_diagnostics": total,
        "errors": sum(1 for d in diagnostics if d["severity"] == "error"),
        "warnings": sum(1 for d in diagnostics if d["severity"] == "warning"),
        "diagnostics": diagnostics[:500],
        "clean": result.returncode == 0,
    }


def run_typecheck(directory: str) -> dict:
    """Run pyright type checker if available."""
    try:
        result = subprocess.run(
            ["pyright", "--outputjson", directory],
            capture_output=True,
            text=True,
            timeout=120,
        )
    except FileNotFoundError:
        return {"error": "pyright not found. Install: npm install -g pyright"}
    except subprocess.TimeoutExpired:
        return {"error": "pyright timed out."}

    try:
        import json

        data = json.loads(result.stdout)
    except (ImportError, json.JSONDecodeError):
        return {"error": f"pyright output error: {result.stderr[:300]}"}

    diagnostics = data.get("generalDiagnostics", [])
    issues = []
    for d in diagnostics[:500]:
        issues.append(
            {
                "file": _fwd(d.get("file", "")),
                "line": d.get("range", {}).get("start", {}).get("line", 0) + 1,
                "severity": d.get("severity", "information").upper(),
                "rule": d.get("rule", ""),
                "message": d.get("message", ""),
            }
        )

    summary = data.get("summary", {})
    return {
        "issues": issues,
        "total": len(issues),
        "errors": summary.get("errorCount", 0),
        "warnings": summary.get("warningCount", 0),
        "informations": summary.get("informationCount", 0),
    }

File: scanner/src/analyzers/format_check.rs
//! Format-checking & type-checking stubs.
//! Rust port of analyzers/format_check.py.
//!
//! These shell out to `ruff` / `ty` / `pyright` — exactly like the Python version.

use std::process::Command;

/// Check formatting with `ruff format --check`.
pub fn check_format(directory: &str) -> serde_json::Value {
    let output = Command::new("ruff")
        .args(["format", "--check", directory])
        .output();

    match output {
        Err(_) => serde_json::json!({"error": "ruff not found. Install: uv tool install ruff"}),
        Ok(out) => {
            let stdout = String::from_utf8_lossy(&out.stdout);
            let stderr = String::from_utf8_lossy(&out.stderr);
            let mut files: Vec<String> = Vec::new();

            for line in stdout.lines() {
                let trimmed = line.trim();
                if !trimmed.is_empty() && std::path::Path::new(trimmed).is_file() {
                    if let Ok(rel) = pathdiff(trimmed, directory) {
                        files.push(fwd(&rel));
                    }
                }
            }
            for line in stderr.lines() {
                let trimmed = line.trim();
                if let Some(rest) = trimmed.strip_prefix("Would reformat:") {
                    let fname = rest.trim();
                    if !fname.is_empty() {
                        files.push(fwd(fname));
                    }
                }
            }

            serde_json::json!({
                "needs_format": files.len(),
                "files": &files[..std::cmp::min(files.len(), 500)],
                "all_formatted": out.status.success(),
            })
        }
    }
}

/// Run `ty` type checker.
pub fn check_types(directory: &str) -> serde_json::Value {
    let output = Command::new("ty")
        .args(["check", "--output-format", "concise", directory])
        .output();

    match output {
        Err(_) => serde_json::json!({"error": "ty not found. Install: uv tool install ty"}),
        Ok(out) => {
            let combined = format!(
                "{}\n{}",
                String::from_utf8_lossy(&out.stdout),
                String::from_utf8_lossy(&out.stderr)
            );
            let mut diagnostics: Vec<serde_json::Value> = Vec::new();
            let mut total_from_summary = 0usize;

            for line in combined.lines() {
                let line = line.trim();
                if line.is_empty() || line.starts_with("Found ") || line.starts_with("info:") {
                    if line.starts_with("Found ") {
                        if let Some(num_str) = line.split_whitespace().nth(1) {
                            if let Ok(n) = num_str.parse::<usize>() {
                                total_from_summary = n;
                            }
                        }
                    }
                    continue;
                }
                if let Some((location, rest)) = line.split_once(": ") {
                    let file_path = location.rsplit_once(':').map_or(location, |(f, _)| f);
                    let severity = if rest.contains("error[") {
                        "error"
                    } else if rest.contains("warning[") {
                        "warning"
                    } else {
                        "info"
                    };
                    diagnostics.push(serde_json::json!({
                        "file": fwd(file_path),
                        "location": location,
                        "message": rest,
                        "severity": severity,
                    }));
                }
            }

            let total = if total_from_summary > 0 {
                total_from_summary
            } else {
                diagnostics.len()
            };
            let errors = diagnostics
                .iter()
                .filter(|d| d["severity"] == "error")
                .count();
            let warnings = diagnostics
                .iter()
                .filter(|d| d["severity"] == "warning")
                .count();

            serde_json::json!({
                "total_diagnostics": total,
                "errors": errors,
                "warnings": warnings,
                "diagnostics": &diagnostics[..std::cmp::min(diagnostics.len(), 500)],
                "clean": out.status.success(),
            })
        }
    }
}

/// Forward-slash normaliser.
fn fwd(p: &str) -> String {
    p.replace('\\', "/")
}

/// Run `pyright` type checker (transpilation of `run_typecheck()` in format_check.py).
pub fn check_types_pyright(directory: &str) -> serde_json::Value {
    let output = Command::new("pyright")
        .args(["--outputjson", directory])
        .output();

    match output {
        Err(_) => serde_json::json!({"error": "pyright not found. Install: npm install -g pyright"}),
        Ok(out) => {
            let stdout = String::from_utf8_lossy(&out.stdout);
            let parsed: Result<serde_json::Value, _> = serde_json::from_str(&stdout);
            match parsed {
                Err(_) => {
                    let stderr = String::from_utf8_lossy(&out.stderr);
                    let snippet: String = stderr.chars().take(300).collect();
                    serde_json::json!({"error": format!("pyright output error: {}", snippet)})
                }
                Ok(data) => {
                    let diagnostics = data.get("generalDiagnostics")
                        .and_then(|v| v.as_array())
                        .cloned()
                        .unwrap_or_default();
                    let mut issues: Vec<serde_json::Value> = Vec::new();
                    for d in diagnostics.iter().take(500) {
                        let file = d.get("file").and_then(|v| v.as_str()).unwrap_or("");
                        let line = d.get("range")
                            .and_then(|r| r.get("start"))
                            .and_then(|s| s.get("line"))
                            .and_then(|l| l.as_u64())
                            .unwrap_or(0) + 1;
                        let severity = d.get("severity")
                            .and_then(|v| v.as_str())
                            .unwrap_or("information")
                            .to_uppercase();
                        let rule = d.get("rule").and_then(|v| v.as_str()).unwrap_or("");
                        let message = d.get("message").and_then(|v| v.as_str()).unwrap_or("");
                        issues.push(serde_json::json!({
                            "file": fwd(file),
                            "line": line,
                            "severity": severity,
                            "rule": rule,
                            "message": message,
                        }));
                    }
                    let summary = data.get("summary").cloned().unwrap_or(serde_json::json!({}));
                    let errors = summary.get("errorCount").and_then(|v| v.as_u64()).unwrap_or(0);
                    let warnings = summary.get("warningCount").and_then(|v| v.as_u64()).unwrap_or(0);
                    let informations = summary.get("informationCount").and_then(|v| v.as_u64()).unwrap_or(0);
                    serde_json::json!({
                        "issues": issues,
                        "total": issues.len(),
                        "errors": errors,
                        "warnings": warnings,
                        "informations": informations,
                    })
                }
            }
        }
    }
}

/// Simple path difference.
fn pathdiff(path: &str, base: &str) -> Result<String, ()> {
    let p = std::path::Path::new(path);
    if let Ok(rel) = p.strip_prefix(base) {
        Ok(rel.to_string_lossy().into_owned())
    } else {
        Ok(path.to_string())
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_fwd() {
        assert_eq!(fwd("a\\b\\c"), "a/b/c");
    }
}
